base_tier_validation_helper/models/tier_definition_helper.py

```diff
@@ -0,0 +1,230 @@
+    @api.constrains('line_ids')
+    def _check_line_ids(self):
+        if self.state == 'done' and \
+                any(not line.tier_definition_id for line in self.line_ids):
+            raise UserError(_('Cannot add a line on Created state!'))
+
+    @api.onchange('model_id')
+    def _onchange_model_id(self):
+        models = self.env['tier.definition']._get_tier_validation_model_names()
+        return {'domain': {'model_id': [('model', 'in', models)]}}
+
+    @api.multi
+    def button_draft(self):
+        self.write({'state': 'draft'})
+
+    @api.multi
+    def button_create_tier(self):
+        if not self.line_ids:
+            raise ValidationError(_('Please add at least one line.'))
+        tier_definition_env = self.env['tier.definition']
+        for rec in self.line_ids:
+            if not rec.tier_definition_id:
+                tier_definition = tier_definition_env.create(
+                    rec._prepare_tier_definition())
+                rec.tier_definition_id = tier_definition
+        self.write({'state': 'done'})
+
+    @api.multi
+    def button_delete_tier(self):
+        self.line_ids.mapped('tier_definition_id').unlink()
+        self.write({'state': 'cancel'})
+
+
+class TierDefinitionHelperLine(models.Model):
+    _name = 'tier.definition.helper.line'
+    _description = 'Tier Definition Helper Line'
+
+    parent_id = fields.Many2one(
+        comodel_name='tier.definition.helper',
+        index=True,
+        required=True,
+        ondelete='cascade'
+    )
+    approval_level = fields.Integer(
+        string='Level',
+        default=1,
+        required=True,
+    )
+    name = fields.Char(
+        string='Name',
+        default='New Tier Definition',
+    )
+    reviewer_id = fields.Many2one(
+        comodel_name="res.users",
+        string="Reviewer",
+    )
+    reviewer_group_id = fields.Many2one(
+        comodel_name="res.groups",
+        string="Reviewer group",
+    )
+    definition_domain = fields.Char(
+        string='Domain',
+        help='Expressions can be any valid python expressions.'
+             '(eg ["|", ["id", "=", 1], ["id", "=", "2"]])',
+    )
+    tier_definition_id = fields.Many2one(
+        comodel_name='tier.definition',
+        string='Tier Definition',
+        readonly=True,
+    )
+
+    def _prepare_tier_definition(self):
+        tier_definition = {
+            'model_id': self.parent_id.model_id.id,
+            'sequence': 100 - self.approval_level,
+            'review_type': self.parent_id.review_type,
+            'definition_type': self.parent_id.definition_type,
+            'notify_on_create': self.parent_id.notify_on_create,
+            'has_comment': self.parent_id.has_comment,
+            'approve_sequence': self.parent_id.approve_sequence,
+        }
+        # add name
+        if self.parent_id.auto_name:
+            tier_definition['name'] = self.parent_id.model_id.name +\
+                ' (' + str(self.approval_level) + ')'
```

Comment:
Please always do not use '+' for sting.
```
name = '%s (%s)' % (var1, var2)
```

Reply:
Anyway, I think auto_name is not needed (let use make name himself is ok).

Reply:
Fixed